## QUESTION 4

In this exercise, we aim to represent various courses on a geographic map. 
Additionally, we want the color of each course to depend on the price.

At first, let's import the dataset and take a look to the variables we are interested in.

In [1]:
import pandas as pd

courses = pd.read_csv("courses_data_processed.tsv", sep='\t')

print(courses[["courseName", "city", "country"]])

                                             courseName     city  \
0              3D Design for Virtual Environments - MSc  Glasgow   
1                          Accounting and Finance - MSc    Leeds   
2     Accounting, Accountability & Financial Managem...   London   
3     Accounting, Financial Management and Digital B...  Reading   
4                                        Addictions MSc   London   
...                                                 ...      ...   
5995                              Materials Engineering    Padua   
5996                          Materials Engineering MSc  Swansea   
5997              Materials Engineering MSc by Research  Swansea   
5998           Materials Engineering with Industry, MSc  Swansea   
5999           Materials for Energy and Environment MSc   London   

             country  
0     United Kingdom  
1     United Kingdom  
2     United Kingdom  
3     United Kingdom  
4     United Kingdom  
...              ...  
5995           Italy  

Now we import some of the packages we will need.

In [2]:
!pip install keplergl
!pip install googlemaps
!pip install gmaps
!pip install geopandas

import pandas as pd
from googlemaps import Client as GoogleMaps
import googlemaps
import gmaps
from keplergl import KeplerGl
import geopandas as gpd

We need 'city' and 'country' to define a new variable, 'Full_Address,' which we will use to import coordinates from Google Maps.

In [3]:
courses['Full_Address'] = courses['city'] + ',' + courses['country']
print(courses['Full_Address'])

0       Glasgow,United Kingdom
1         Leeds,United Kingdom
2        London,United Kingdom
3       Reading,United Kingdom
4        London,United Kingdom
                 ...          
5995               Padua,Italy
5996    Swansea,United Kingdom
5997    Swansea,United Kingdom
5998    Swansea,United Kingdom
5999     London,United Kingdom
Name: Full_Address, Length: 6000, dtype: object


Now let's insert the API key obtained by signing up for Google Developer Console, to obtain latitude and longitude coordinates from 'Full_Address'.

In [4]:
gmaps = googlemaps.Client(key='AIzaSyBArdb62x51edUdvhgJszeJ_LXFU8jibfY')

Let's just pass the 'Full_address' column to the google API.

In [5]:
courses1= courses.iloc[:,-1:].copy()
courses1.head()

Full_Address
0  Glasgow,United Kingdom
1    Leeds,United Kingdom
2   London,United Kingdom
3  Reading,United Kingdom
4   London,United Kingdom

Now we create two empty columns that will be filled with the latitude and longitude coordinates.

In [6]:
courses1['long'] = ""
courses1['lat'] = ""

Using the Google Maps Geocoding API to retrieve latitude and longitude coordinates for addresses stored in the 'Full_Address' column of the 'courses1' DataFrame.

In [7]:
for x in range(len(courses1)): #iterate over the rows of the 'courses1'
    geocode_result = gmaps.geocode(courses1['Full_Address'][x]) #send a geocoding request to the Google Maps Geocoding API for the address in the 'Full_Address' column of the current row
    if geocode_result: #check if the geocoding request was successful
        #latitude and longitude coordinates are extracted from the API response and assigned to the 'lat' and 'long' columns of the 'courses1'
        courses1['lat'][x] = geocode_result[0]['geometry']['location']['lat']
        courses1['long'][x] = geocode_result[0]['geometry']['location']['lng']
    else:
        #if we don't have any result it sets the 'lat' and 'long' columns for the current row to empty strings
        courses1['lat'][x] = "" 
        courses1['long'][x] = ""

Insert the new variables in the original dataset.

In [8]:
courses['Lat']=courses1['lat']
courses['Lon']= courses1['long']
courses.head()

courseName  \
0           3D Design for Virtual Environments - MSc   
1                       Accounting and Finance - MSc   
2  Accounting, Accountability & Financial Managem...   
3  Accounting, Financial Management and Digital B...   
4                                     Addictions MSc   

                  universityName  \
0  Glasgow Caledonian University   
1            University of Leeds   
2          King’s College London   
3          University of Reading   
4          King’s College London   

                                         facultyName         isItFullTime  \
0        School of Engineering and Built Environment            Full time   
1                   Leeds University Business School            Full time   
2                             King’s Business School            Full time   
3                             Henley Business School            Full time   
4  Institute of Psychiatry, Psychology and Neuros...  Full time&Part time   

                                         description  startDate     fees  \
0  3D visualisation and animation play a role in ...  September      NaN   
1  Businesses and governments rely on sound finan...  September  34750.0   
2  Our Accounting, Accountability & Financial Man...  September      NaN   
3  Embark on a professional accounting career wit...  September      NaN   
4  Join us for an online session for prospective ...  September      NaN   

  modality          duration     city         country administration  \
0      MSc  1 year full-time  Glasgow  United Kingdom      On Campus   
1      MSc  1 year full time    Leeds  United Kingdom      On Campus   
2      MSc         1 year FT   London  United Kingdom      On Campus   
3      MSc  1 year full time  Reading  United Kingdom      On Campus   
4      MSc       One year FT   London  United Kingdom      On Campus   

                                                 url  \
0  https://www.findamasters.com/masters-degrees/c...   
1  https://www.findamasters.com/masters-degrees/c...   
2  https://www.findamasters.com/masters-degrees/c...   
3  https://www.findamasters.com/masters-degrees/c...   
4  https://www.findamasters.com/masters-degrees/c...   

                                ProcessedDescription currency  fees (EUR)  \
0  3d visualis anim play role mani area popular m...      NaN         NaN   
1  busi govern reli sound financi knowledg underp...        £    39869.21   
2  account account  financi manag msc cours provi...      NaN         NaN   
3     embark profession account career academ ground      NaN         NaN   
4  join us onlin session prospect student find ms...      NaN         NaN   

             Full_Address        Lat       Lon  
0  Glasgow,United Kingdom   55.86167 -4.258335  
1    Leeds,United Kingdom  53.800755 -1.549077  
2   London,United Kingdom  51.507218 -0.127586  
3  Reading,United Kingdom   51.45512 -0.978747  
4   London,United Kingdom  51.507218 -0.127586

At this point, we have the coordinates for the city and country of each course. 
Now, we want to be able to visualize information about the cost of the courses on the map.

In [9]:
#Drop the null values
courses = courses.dropna(subset=["fees (EUR)"]) 
print(courses["fees (EUR)"])

1       39869.21
5       35566.77
7       15000.00
8       15000.00
9       32985.31
          ...   
5982     3203.07
5983    29900.00
5984    29900.00
5992    16177.15
5999    16177.15
Name: fees (EUR), Length: 1228, dtype: float64


Now we have 1228 courses.

Let's divide the variable into price ranges.

In [10]:
#See what's the maximum price
print(max(courses['fees (EUR)']))

114160191.6


In [11]:
#Define the ranges and create a new variable 
intervals=[0, 1000, 10000, 100000, 1000000, max(courses['fees (EUR)'])]
labels = ['<1k', '1k-10k', '10k-100k', '100k-1M', '>1M']
courses['fees_interval'] = pd.cut(courses['fees (EUR)'], bins=intervals, labels=labels)

#Let's see how many courses we have for each range
print(courses['fees_interval'].value_counts())

10k-100k    936
1k-10k      148
<1k         118
100k-1M       9
>1M           3
Name: fees_interval, dtype: int64


Let's see our first option for the map. It is interactive so you must run the code to see it.

In [26]:
!pip install folium
!pip install folium --upgrade

import folium
from folium.plugins import MarkerCluster

# Create a map centered on an initial location.
m = folium.Map(location=[courses['Lat'].mean(), courses['Lon'].mean()], zoom_start=10)

# Create a marker cluster.
marker_cluster = MarkerCluster().add_to(m)

# Define colors and default color.
colors = {'<1k': 'blue', '1k-10k': 'green', '10k-100k': 'orange', '100k-1M': 'red', '>1M': 'purple'}
default_color = 'gray'  # Default color for values NaN

# Iterate through the DataFrame and add markers to the cluster
for index, row in courses.iterrows():
    fee_interval = row['fees_interval'] if not pd.isna(row['fees_interval']) else None
    color = colors.get(fee_interval, default_color)

    # Add the marker to the cluster with the appropriate color, and the course name as a label.
    folium.Marker([row['Lat'], row['Lon']], popup=row['Full_Address'], icon=folium.Icon(color=color), tooltip=row['courseName']).add_to(marker_cluster)

# Add legend
legend_html = '''
     <div style="position: fixed; 
                 bottom: 50px; left: 50px; width: 150px; height: 180px; 
                 border:2px solid grey; z-index:9999; font-size:14px;
                 background-color:white; opacity: .85;">
     &nbsp; Fees ranges <br>
     &nbsp; <1k <i style="background: blue; width: 18px; height: 18px; border-radius: 50%; display: inline-block;"></i><br>
     &nbsp; 1k-10k <i style="background: green; width: 18px; height: 18px; border-radius: 50%; display: inline-block;"></i><br>
     &nbsp; 10k-100k <i style="background: orange; width: 18px; height: 18px; border-radius: 50%; display: inline-block;"></i><br>
     &nbsp; 100k-1M <i style="background: red; width: 18px; height: 18px; border-radius: 50%; display: inline-block;"></i><br>
     &nbsp; >1M <i style="background: purple; width: 18px; height: 18px; border-radius: 50%; display: inline-block;"></i><br>
      </div>
     '''

m.get_root().html.add_child(folium.Element(legend_html))

# Display the map
m


The second option is:

In [27]:
import plotly.express as px
import geopandas as gpd
import pandas as pd

#variables which will be visible for each course on the map
hover_data_columns = ['courseName', 'fees_interval', 'city', 'country']

#dot's color
color_order = ['blue', 'green', 'orange', 'red', 'purple'] 

#define the points: location and color
fig = px.scatter_geo(courses,
                     lat=courses['Lat'], 
                     lon=courses['Lon'], 
                     hover_name="courseName", 
                     color="fees_interval",
                     category_orders={"fees_interval": labels},
                     color_discrete_sequence=color_order,
                     projection="natural earth",
                     custom_data=courses[hover_data_columns])

#add the legend
fig.update_layout(legend_title_text='Fees Interval')

#customize the hover information that is displayed when you hover over a data point on the Plotly scatter plot.
fig.update_traces(hovertemplate="<br>".join(["%{customdata[" + str(i) + "]}" for i in range(len(hover_data_columns))]))

#show the figure
fig.show()
